### PR TITLE
docs: propose a Shuffle operator

### PR DESCRIPTION
**Pull request by pingyu**
### What problem does this PR solve? 
Design document of Shuffle Executor proposed and implemented in #14238 

### What is changed and how it works?
New design document and related diagrams.

### Check List 

Tests 
 - No code

### CODE REVIEW

File: docs/design/2020-01-10-shuffle-executor.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+The implementation consists of 3 parts:
+1. Physical Planning.
+2. Building Executors.
+3. Executing.
+
+### Planning
+
+We find the very position and very moment for "shuffle" in Physical Planning procedure.
+
+The very position is on top of the operator which can be "shuffled", i.e. Window operator so far.
```
Comment (zz-jason) [inline docs/design/2020-01-10-shuffle-executor.md]: IMHO, shuffle is used to redistribute data: it changes the data property from 1 distribution to another distribution. In a multi-thread execution model, the operator which can change the data distribution can be `group by`, `join`, etc.

In order to correctly handle the `shuffle` operator, we should consider the distribution property in the physical optimization phase.

File: docs/design/2020-01-10-shuffle-executor.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+| id                         | count    | task      | operator info                                              |
++----------------------------+----------+-----------+------------------------------------------------------------+
+| Projection_7               | 10000.00 | root      | test.t.j, Column#5                                         |
+| └─Shuffle_12               | 10000.00 | root      | execution info: concurrency:4, data source:TableReader_10  |
+|   └─Window_8               | 10000.00 | root      | sum(cast(test.t.i))->Column#5 over(partition by test.t.j)  |
+|     └─Sort_11              | 10000.00 | root      | test.t.j:asc                                               |
+|       └─TableReader_10     | 10000.00 | root      | data:TableScan_9                                           |
+|         └─TableScan_9      | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
++----------------------------+----------+-----------+------------------------------------------------------------+
+```
```
Comment (eurekaka) [inline docs/design/2020-01-10-shuffle-executor.md]: This new plan is pretty strange. Actually in execution, the output of `Shuffle` should be the input of `Window`, so a more intuitive plan from my perspective would be like `TableScan -> TableReader -> Sort -> Shuffle -> Window -> Merge -> Projection`, as you depicted in the diagram above.

Comment (pingyu) [reply docs/design/2020-01-10-shuffle-executor.md]: I agree. I will change this document and related codes in the next couple of days.

File: docs/design/2020-01-10-shuffle-executor.md (hunk 1 above)
Comment (pingyu) [reply docs/design/2020-01-10-shuffle-executor.md]: It means that we should do "shuffle" according to the physical property, e.g. hash on unsorted data, streaming on sorted data. Isn't it ?

File: docs/design/2020-01-10-shuffle-executor.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+The very moment, means all of the following conditions should be met:
+
+- The corresponding parallelism switch variable is on. As to Window operator, the value of variable `tidb_window_concurreny` should be more than 1.
+- The NDV (Number of Dinstict Values) of `PARTITION BY` columns should be more than 1.
+- The parallelism should be effective enough. As to Window operator, when the child of Window operator meets the sorted property requirement (i.e. enforced sorting is not necessary), the parallelism is not effective enough. See "Benchmarks" section for detail.
+
+(_Notice that Window operator implements grouping in a "streaming" manner. So if data source is not sorted on the `PARTITION BY` columns, a Sorting would be enforced. Moreover, `SORT BY` clause in Window operator requires sorted property of data source too._)
+
+When the very position and very moment is found, we inserts a new physical plan `PhysicalShuffle`, which represents the executing plan of Shuffle Executor.
```
Comment (qw4990) [inline docs/design/2020-01-10-shuffle-executor.md]: Is there a `LogicalShuffle` and should we let the planner decide if we use `Shuffle`?

File: docs/design/2020-01-10-shuffle-executor.md (hunk 1 above)
Comment (pingyu) [reply docs/design/2020-01-10-shuffle-executor.md]: @zz-jason 
Comment addressed, PTAL~

File: docs/design/2020-01-10-shuffle-executor.md (hunk 2 above)
Comment (pingyu) [reply docs/design/2020-01-10-shuffle-executor.md]: @eurekaka 
Comment addressed, PTAL~

File: docs/design/2020-01-10-shuffle-executor.md (hunk 3 above)
Comment (pingyu) [reply docs/design/2020-01-10-shuffle-executor.md]: By now whether using `Shuffle` or not, lies on the following conditions. The 2 & 3 should be check in physical planing, I think.
1. The corresponding parallelism switch variable is on. As to Window operator, the value of variable `tidb_window_concurreny` should be more than 1.
2. The NDV (Number of Dinstict Values) of `PARTITION BY` columns should be more than 1.
3. The parallelism should be effective enough. As to Window operator, when the child of Window operator meets the sorted property requirement (i.e. enforced sorting is not necessary), the parallelism is not effective enough. See "Benchmarks" section for detail.

**Comment by zz-jason** [comment]
can we close this PR since it hasn't been updated for nearly 1 year?

**Comment by zz-jason** [comment]
I'm going to close this PR due to it's stale, feel free to reopen it if you are going to update this PR in the future.

**Pull request closed without merging**